presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseToMap.java

```diff
@@ -0,0 +1,278 @@
+                    ((ConstantExpression) key).getValue() == null ||
+                    (keys.size() > 0 && !keys.get(0).getType().equals(key.getType()))) {
+                return false;
+            }
+
+            if (keySet.add(key)) {
+                // We allow all same type only
+                if (values.size() > 0 && !values.get(0).getType().equals(value.getType())) {
+                    return false;
+                }
+
+                keys.add(key);
+                values.add(value);
+            }
+
+            return true;
+        }
+
+        @Override
+        public RowExpression rewriteSpecialForm(SpecialFormExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getForm() != SWITCH) {
+                return rewriteRowExpression(node, context, treeRewriter);
+            }
+
+            // by construction we should have at least one WHEN and ELSE is always added if misseing
+            int numArgs = node.getArguments().size();
+            RowExpression lastArg = node.getArguments().get(numArgs - 1);
+
+            checkState(numArgs >= 2);
+            checkState(!(lastArg instanceof SpecialFormExpression && ((SpecialFormExpression) lastArg).getForm().equals(WHEN)));
+
+            if (!(lastArg instanceof ConstantExpression)) {
+                return node;
+            }
+
+            RowExpression firstArg = node.getArguments().get(0);
+            Set<RowExpression> keySet = new HashSet<>();
+            List<RowExpression> whens = new ArrayList<RowExpression>(node.getArguments().size());
+            List<RowExpression> thens = new ArrayList<RowExpression>(node.getArguments().size());
+            RowExpression checkExpr;
+            int start;
+
+            if (!(firstArg instanceof SpecialFormExpression && ((SpecialFormExpression) firstArg).getForm().equals(WHEN))) {
+                if (firstArg.equals(constant(true, BOOLEAN))) {
+                    // We generate weird CASE (true) WHEN p1 THEN v1 etc. for non-searched case
+                    // So drop the true
+                    checkExpr = null;
+                }
+                else {
+                    checkExpr = firstArg;
+                }
+
+                start = 1;
+            }
+            else {
+                checkExpr = null;
+                start = 0;
+            }
+
+            for (int i = start; i < numArgs - 1; i++) {
+                RowExpression whenClause = node.getArguments().get(i);
+                RowExpression value = whenClause.getChildren().get(1);
+
+                if (!(value instanceof ConstantExpression)) {
+                    // THEN is not a constant
+                    return node;
+                }
+
+                RowExpression when = whenClause.getChildren().get(0);
+                RowExpression key;
+                RowExpression curCheck;
+
+                if (when instanceof ConstantExpression) {
```

Comment:
maybe split up the three handled cases into separate functions to make it more readable + add an example for each of the cases

Reply:
too much state copied around but yeah probably a good idea